In [2]:
#from sklearn.model_selection import train_test_split
#from sklearn.model_selection import ParameterGrid

from sklearn.linear_model import LinearRegression
from sklearn.linear_model import ElasticNet

from sklearn.ensemble import GradientBoostingRegressor
from sklearn.ensemble import GradientBoostingClassifier
from sklearn.ensemble import RandomForestRegressor
from sklearn.ensemble import IsolationForest

from sklearn.tree import DecisionTreeClassifier
from sklearn.tree import DecisionTreeRegressor, plot_tree

#from sklearn_pandas import DataFrameMapper
from sklearn.compose import ColumnTransformer

from sklearn.impute import SimpleImputer

from sklearn.preprocessing import StandardScaler
from sklearn.preprocessing import MinMaxScaler
from sklearn.preprocessing import OneHotEncoder

from sklearn.pipeline import make_pipeline, Pipeline

from sklearn.metrics import mean_squared_error
from sklearn.metrics import mean_absolute_percentage_error
#from sklearn.metrics import PredictionErrorDisplay

from sklearn.decomposition import PCA

#from sklearn.model_selection import cross_val_score, GridSearchCV, KFold, RandomizedSearchCV, train_test_split

from tqdm import tqdm

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import scipy.stats
import datetime
import seaborn as sns
sns.set_palette('muted')
sns.set_color_codes('muted')
sns.set_style('white')

import warnings
warnings.filterwarnings('ignore')

from statsmodels.graphics.mosaicplot import mosaic
pd.set_option('display.max_columns', 70)

In [3]:
pd.set_option('display.max_colwidth', None)

In [4]:
from mlxtend.frequent_patterns import apriori, association_rules

In [5]:
#pip install mlxtend

In [6]:
path = 'C:/Users/elysi/Tele2_project/data/pre_model_state/df_ARL_combined_share_percentiles_yandex_1299_features_final_binary_df.csv'
df = pd.read_csv(path)

In [7]:
#df = pd.read_csv('C:/Users/elysi/Tele2_project/data/pre_model_state/df_ARL_combined_share_percentiles_no_osm_981_features_22_04_2024.csv')

In [8]:
df = df.replace(np.nan, 0)

In [9]:
df.set_index('h3_9', inplace=True)

In [10]:
binary_df = df.astype(int)

In [11]:
binary_df.shape

(1335, 1303)

In [12]:
binary_df.head(3)

total_visits_0_20_prcntl  total_visits_20_40_prcntl  \
h3_9                                                                   
8911060006bffff                         1                          0   
89110600077ffff                         1                          0   
89110600113ffff                         0                          1   

                 total_visits_40_60_prcntl  total_visits_60_80_prcntl  \
h3_9                                                                    
8911060006bffff                          0                          0   
89110600077ffff                          0                          0   
89110600113ffff                          0                          0   

                 total_visits_80_100_prcntl  car_owners_total_0_20_prcntl  \
h3_9                                                                        
8911060006bffff                           0                             1   
89110600077ffff                           0                             1   
89110600113ffff                           0                             0   

                 car_owners_total_20_40_prcntl  car_owners_total_40_60_prcntl  \
h3_9                                                                            
8911060006bffff                              0                              0   
89110600077ffff                              0                              0   
89110600113ffff                              1                              0   

                 car_owners_total_60_80_prcntl  \
h3_9                                             
8911060006bffff                              0   
89110600077ffff                              0   
89110600113ffff                              0   

                 car_owners_total_80_100_prcntl  \
h3_9                                              
8911060006bffff                               0   
89110600077ffff                               0   
89110600113ffff                               0   

                 carsharing_user_total_0_20_prcntl  \
h3_9                                                 
8911060006bffff                                  1   
89110600077ffff                                  0   
89110600113ffff                                  0   

                 carsharing_user_total_20_40_prcntl  \
h3_9                                                  
8911060006bffff                                   0   
89110600077ffff                                   1   
89110600113ffff                                   1   

                 carsharing_user_total_40_60_prcntl  \
h3_9                                                  
8911060006bffff                                   0   
89110600077ffff                                   0   
89110600113ffff                                   0   

                 carsharing_user_total_60_80_prcntl  \
h3_9                                                  
8911060006bffff                                   0   
89110600077ffff                                   0   
89110600113ffff                                   0   

                 carsharing_user_total_80_100_prcntl  \
h3_9                                                   
8911060006bffff                                    0   
89110600077ffff                                    0   
89110600113ffff                                    0   

                 not_citizen_rf_total_0_20_prcntl  \
h3_9                                                
8911060006bffff                                 1   
89110600077ffff                                 1   
89110600113ffff                                 0   

                 not_citizen_rf_total_20_40_prcntl  \
h3_9                                                 
8911060006bffff                                  0   
89110600077ffff                                  0   
89110600113ffff                                  1   

                 not_citizen_rf_total_40_60_prcntl  \
h3_9                

## Apriori with 20 percentile step for each column

**Other algorytms like FP-Grows work the same, we can use it for optimized runs. In theory it provided opportunity for 3 rules combination. Nevertheless, currently we work we pairs of rules which apriori algorythm can handle**

In [13]:
%%time

frequent_itemsets = apriori(binary_df, min_support=0.1, use_colnames=True, verbose=1, max_len = 2)

rules = association_rules(frequent_itemsets, metric="confidence", min_threshold=0.1)
rules_sorted = rules.sort_values(by='lift', ascending=False)
rules_sorted.head(10)

D:\Anaconda\lib\site-packages\mlxtend\frequent_patterns\fpcommon.py:109: DeprecationWarning: DataFrames with non-bool types result in worse computationalperformance and their support might be discontinued in the future.Please use a DataFrame with bool type
  warnings.warn(


Processing 1027182 combinations | Sampling itemset size 2
CPU times: total: 6.53 s
Wall time: 16 s


antecedents  \
23737               (category_free_parking_count_2_hn_20_40_prcntl)   
23736            (subcategory_free_parking_count_2_hn_20_40_prcntl)   
24071  (subcategory_industrial_enterprises_count_2_hn_20_40_prcntl)   
24070     (category_industrial_enterprises_count_2_hn_20_40_prcntl)   
23437                  (subcategory_hotels_count_2_hn_20_40_prcntl)   
23436                     (category_hotels_count_2_hn_20_40_prcntl)   
23861            (subcategory_paid_parking_count_2_hn_20_40_prcntl)   
23860               (category_paid_parking_count_2_hn_20_40_prcntl)   
22472                  (category_hospitals_count_2_hn_40_60_prcntl)   
22473               (subcategory_hospitals_count_2_hn_40_60_prcntl)   

                                                        consequents  \
23737            (subcategory_free_parking_count_2_hn_20_40_prcntl)   
23736               (category_free_parking_count_2_hn_20_40_prcntl)   
24071     (category_industrial_enterprises_count_2_hn_20_40_prcntl)   
24070  (subcategory_industrial_enterprises_count_2_hn_20_40_prcntl)   
23437                     (category_hotels_count_2_hn_20_40_prcntl)   
23436                  (subcategory_hotels_count_2_hn_20_40_prcntl)   
23861               (category_paid_parking_count_2_hn_20_40_prcntl)   
23860            (subcategory_paid_parking_count_2_hn_20_40_prcntl)   
22472               (subcategory_hospitals_count_2_hn_40_60_prcntl)   
22473                  (category_hospitals_count_2_hn_40_60_prcntl)   

       antecedent support  consequent support   support  confidence      lift  \
23737            0.104869            0.104869  0.104869         1.0  9.535714   
23736            0.104869            0.104869  0.104869         1.0  9.535714   
24071            0.110112            0.110112  0.110112         1.0  9.081633   
24070            0.110112            0.110112  0.110112         1.0  9.081633   
23437            0.129588            0.129588  0.129588         1.0  7.716763   
23436            0.129588            0.129588  0.129588         1.0  7.716763   
23861            0.133333            0.133333  0.133333         1.0  7.500000   
23860            0.133333            0.133333  0.133333         1.0  7.500000   
22472            0.147566            0.147566  0.147566         1.0  6.776650   
22473            0.147566            0.147566  0.147566         1.0  6.776650   

       leverage  conviction  zhangs_metric  
23737  0.093871         inf            1.0  
23736  0.093871         inf            1.0  
24071  0.097988         inf            1.0  
24070  0.097988         inf            1.0  
23437  0.112795         inf            1.0  
23436  0.112795         inf            1.0  
23861  0.115556         inf            1.0  
23860  0.115556         inf            1.0  
22472  0.125790         inf            1.0  
22473  0.125790         inf            1.0

In [14]:
rules_sorted.shape

(24832, 10)

In [15]:
path = 'C:/Users/elysi/Tele2_project/data/pre_model_state/ARL_salary_60_80_100.csv'
rules_sorted.to_csv(path, index=False)

**we have such a small tresholds for support and confidence because of the dataset form. Since it based on 5 equal percentile range groups, there is no features with high support rate** 

**Check specific column: fit_clubs**

In [16]:
interesting_columns =  [#'total_fit_clubs_neighbour_2_hn_0_20_prcntl',
       #'total_fit_clubs_neighbour_2_hn_20_40_prcntl',
       'total_fit_clubs_neighbour_2_hn_60_80_prcntl',
       'total_fit_clubs_neighbour_2_hn_80_100_prcntl',
       'total_fit_clubs_binary']

In [17]:
filtered_rules = rules[
    #(rules['antecedents'].apply(lambda x: any(item in x for item in interesting_columns_super_rich)) & 
     #rules['consequents'].apply(lambda x: not any("salary" in item for item in x))) |
    
    (rules['consequents'].apply(lambda x: any(item in x for item in interesting_columns)) & 
     rules['antecedents'].apply(lambda x: not any("fit" in item for item in x)))
]

# Sort the filtered rules by the lift metric in descending order
filtered_rules_sorted = filtered_rules.sort_values(by='lift', ascending=False)
len(filtered_rules_sorted)

6

In [18]:
usefull_column = ['antecedents', 'consequents',
                 'support', 'confidence', 'lift']

In [19]:
filtered_rules_sorted[usefull_column]

antecedents  \
21340                       (total_subcategory_museums_neighbour_4_hn_binary)   
21336                       (total_subcategory_cinemas_neighbour_4_hn_binary)   
12279                                (sprtswr_ftwr_inter_unique_60_80_prcntl)   
21343  (total_subcategory_parks_of_culture_and_leisure_neighbour_4_hn_binary)   
21338              (total_subcategory_shopping_centers_neighbour_4_hn_binary)   
11382                                (elite_sports_inter_unique_60_80_prcntl)   

                                         consequents   support  confidence  \
21340  (total_fit_clubs_neighbour_2_hn_60_80_prcntl)  0.152060    0.249386   
21336  (total_fit_clubs_neighbour_2_hn_60_80_prcntl)  0.180524    0.218297   
12279  (total_fit_clubs_neighbour_2_hn_60_80_prcntl)  0.100375    0.211690   
21343  (total_fit_clubs_neighbour_2_hn_60_80_prcntl)  0.155805    0.210740   
21338  (total_fit_clubs_neighbour_2_hn_60_80_prcntl)  0.199251    0.207812   
11382  (total_fit_clubs_neighbour_2_hn_60_80_prcntl)  0.113858    0.207367   

           lift  
21340  1.237658  
21336  1.083370  
12279  1.050582  
21343  1.045864  
21338  1.031337  
11382  1.029126

In [20]:
def contains_stop_word(items, stop_words):
    return any(stop_word in item for stop_word in stop_words for item in items)

# Define a function to check if any interesting column is in the consequent
def contains_interesting_column(items, interesting_columns):
    return any(interesting_column in item for interesting_column in interesting_columns for item in items)

In [21]:
interesting_columns =  ['total_fit_clubs_neighbour_2_hn_0_20_prcntl',
       'total_fit_clubs_neighbour_2_hn_20_40_prcntl']

In [22]:
stop_words = ['society', 'computers', 'services']

In [23]:
# Filter the rules
filtered_rules = rules[
    #(rules['antecedents'].apply(lambda x: contains_interesting_column(x, interesting_columns_super_rich)) & 
     #rules['consequents'].apply(lambda x: not contains_stop_word(x, stop_words))) |

    (rules['consequents'].apply(lambda x: contains_interesting_column(x, interesting_columns)) & 
     rules['antecedents'].apply(lambda x: not contains_stop_word(x, stop_words)))
]

# Sort the filtered rules by the lift metric in descending order
filtered_rules_sorted = filtered_rules.sort_values(by='lift', ascending=False)
len(filtered_rules_sorted)

92

In [24]:
usefull_column = ['antecedents', 'consequents',
                 'support', 'confidence', 'lift']

In [25]:
filtered_rules_sorted[usefull_column].head(10)

antecedents  \
21220               (category_culture_and_art_count_2_hn_0_20_prcntl)   
21231         (category_science_and_education_count_2_hn_0_20_prcntl)   
21241  (category_recreation_and_entertainment_count_2_hn_0_20_prcntl)   
21224                        (category_stores_count_2_hn_0_20_prcntl)   
21204                      (category_business_count_2_hn_0_20_prcntl)   
21260                       (category_tourism_count_2_hn_0_20_prcntl)   
21210             (category_home_and_interior_count_2_hn_0_20_prcntl)   
21218             (category_beauty_and_health_count_2_hn_0_20_prcntl)   
21226         (category_medicine_and_pharmacy_count_2_hn_0_20_prcntl)   
21254            (category_goods_for_children_count_2_hn_0_20_prcntl)   

                                        consequents   support  confidence  \
21220  (total_fit_clubs_neighbour_2_hn_0_20_prcntl)  0.146067    0.709091   
21231  (total_fit_clubs_neighbour_2_hn_0_20_prcntl)  0.144569    0.704380   
21241  (total_fit_clubs_neighbour_2_hn_0_20_prcntl)  0.146067    0.691489   
21224  (total_fit_clubs_neighbour_2_hn_0_20_prcntl)  0.137828    0.678967   
21204  (total_fit_clubs_neighbour_2_hn_0_20_prcntl)  0.134831    0.674157   
21260  (total_fit_clubs_neighbour_2_hn_0_20_prcntl)  0.135581    0.672862   
21210  (total_fit_clubs_neighbour_2_hn_0_20_prcntl)  0.137828    0.666667   
21218  (total_fit_clubs_neighbour_2_hn_0_20_prcntl)  0.132584    0.660448   
21226  (total_fit_clubs_neighbour_2_hn_0_20_prcntl)  0.133333    0.659259   
21254  (total_fit_clubs_neighbour_2_hn_0_20_prcntl)  0.133333    0.652015   

           lift  
21220  2.094328  
21231  2.080413  
21241  2.042341  
21224  2.005355  
21204  1.991150  
21260  1.987326  
21210  1.969027  
21218  1.950659  
21226  1.947148  
21254  1.925751

**Check specific column: salary**

In [26]:
interesting_columns_rich = [
 'salary_5_unique_share_0_20_prcntl',
 'salary_5_unique_share_20_40_prcntl',
 #'salary_5_unique_share_40_60_prcntl',
 'salary_5_unique_share_60_80_prcntl',
 'salary_5_unique_share_80_100_prcntl']

In [27]:
interesting_columns_super_rich = [
 'salary_5_unique_share_60_80_prcntl',
 'salary_5_unique_share_80_100_prcntl']

Main filter to research rules:
1. exclude data leak rules (f.y. salary to salary feature)
2. search any interested feature from specified group (f.i. interesting_columns_super_rich)

In [28]:
interesting_columns =  [
 'salary_4_unique_share_0_20_prcntl',
 'salary_4_unique_share_20_40_prcntl',
 #'salary_4_unique_share_40_60_prcntl',
 'salary_4_unique_share_60_80_prcntl',
 'salary_4_unique_share_80_100_prcntl'
 'salary_5_unique_share_0_20_prcntl',
 'salary_5_unique_share_20_40_prcntl',
 #'salary_5_unique_share_40_60_prcntl',
 'salary_5_unique_share_60_80_prcntl',
 'salary_5_unique_share_80_100_prcntl']

In [29]:
stop_words = ['salary']

In [30]:
# Filter the rules
filtered_rules = rules[
    #(rules['antecedents'].apply(lambda x: contains_interesting_column(x, interesting_columns_super_rich)) & 
     #rules['consequents'].apply(lambda x: not contains_stop_word(x, stop_words))) |

    (rules['consequents'].apply(lambda x: contains_interesting_column(x, interesting_columns_rich)) & 
     rules['antecedents'].apply(lambda x: not contains_stop_word(x, stop_words)))
]

# Sort the filtered rules by the lift metric in descending order
filtered_rules_sorted = filtered_rules.sort_values(by='lift', ascending=False)
len(filtered_rules_sorted)

18

In [31]:
usefull_column = ['antecedents', 'consequents',
                 'support', 'confidence', 'lift']

In [32]:
filtered_rules_sorted[filtered_rules_sorted['lift'] > 1][usefull_column].head(8)

antecedents  \
17406                       (total_subcategory_museums_neighbour_4_hn_binary)   
14493                                            (age_20_unique_60_80_prcntl)   
17408  (total_subcategory_parks_of_culture_and_leisure_neighbour_4_hn_binary)   
12165                                (sprtswr_ftwr_inter_unique_60_80_prcntl)   
14491                                            (age_20_unique_60_80_prcntl)   
10902                                (elite_sports_inter_unique_60_80_prcntl)   
17382              (total_subcategory_shopping_centers_neighbour_4_hn_binary)   
14489                                            (age_20_unique_60_80_prcntl)   

                                consequents   support  confidence      lift  
17406  (salary_5_unique_share_60_80_prcntl)  0.137079    0.224816  1.145531  
14493  (salary_5_unique_share_60_80_prcntl)  0.107865    0.216541  1.103369  
17408  (salary_5_unique_share_60_80_prcntl)  0.155056    0.209726  1.068644  
12165   (salary_5_unique_share_0_20_prcntl)  0.101873    0.214850  1.062313  
14491  (salary_5_unique_share_20_40_prcntl)  0.104120    0.209023  1.053000  
10902  (salary_5_unique_share_60_80_prcntl)  0.111610    0.203274  1.035767  
17382   (salary_5_unique_share_0_20_prcntl)  0.195506    0.203906  1.008203  
14489   (salary_5_unique_share_0_20_prcntl)  0.101124    0.203008  1.003759

**Check specific column: total visits**

In [33]:
binary_df.filter(regex = 'evening_share').columns

Index(['evening_share_0_20_prcntl', 'evening_share_20_40_prcntl',
       'evening_share_40_60_prcntl', 'evening_share_60_80_prcntl',
       'evening_share_80_100_prcntl'],
      dtype='object')

In [34]:
interesting_columns =  ['evening_share_0_20_prcntl', 'evening_share_20_40_prcntl',
       'evening_share_60_80_prcntl',
       'evening_share_80_100_prcntl']

In [35]:
stop_words = ['evening']

In [36]:
# Filter the rules
filtered_rules = rules[
    #(rules['antecedents'].apply(lambda x: contains_interesting_column(x, interesting_columns_super_rich)) & 
     #rules['consequents'].apply(lambda x: not contains_stop_word(x, stop_words))) |

    (rules['consequents'].apply(lambda x: contains_interesting_column(x, interesting_columns)) & 
     rules['antecedents'].apply(lambda x: not contains_stop_word(x, stop_words)))
]

# Sort the filtered rules by the lift metric in descending order
filtered_rules_sorted = filtered_rules.sort_values(by='lift', ascending=False)
len(filtered_rules_sorted)

37

In [37]:
usefull_column = ['antecedents', 'consequents',
                 'support', 'confidence', 'lift']

In [38]:
filtered_rules_sorted[filtered_rules_sorted['lift'] > 1][usefull_column].head(10)

antecedents  \
20302                     (Friday_share_0_20_prcntl)   
20190                  (Stay9_15h_share_0_20_prcntl)   
17435                 (avg_duration_hrs_0_20_prcntl)   
20452                   (Thursday_share_0_20_prcntl)   
20593                      (night_share_0_20_prcntl)   
19973             (weekday_visits_share_0_20_prcntl)   
20594  (total_population_neighbour_2_hn_0_20_prcntl)   
20230                 (Stay15_20h_share_0_20_prcntl)   
17494              (median_duration_hrs_0_20_prcntl)   
18369                        (Stay9_15h_0_20_prcntl)   

                       consequents   support  confidence      lift  
20302  (evening_share_0_20_prcntl)  0.137828    0.689139  3.445693  
20190  (evening_share_0_20_prcntl)  0.124345    0.621723  3.108614  
17435  (evening_share_0_20_prcntl)  0.117603    0.588015  2.940075  
20452  (evening_share_0_20_prcntl)  0.116105    0.580524  2.902622  
20593  (evening_share_0_20_prcntl)  0.112360    0.561798  2.808989  
19973  (evening_share_0_20_prcntl)  0.110112    0.550562  2.752809  
20594  (evening_share_0_20_prcntl)  0.107865    0.539326  2.696629  
20230  (evening_share_0_20_prcntl)  0.107116    0.535581  2.677903  
17494  (evening_share_0_20_prcntl)  0.107116    0.533582  2.667910  
18369  (evening_share_0_20_prcntl)  0.107865    0.525547  2.627737

In [251]:
binary_df.filter(regex = 'morning_share').columns

Index(['morning_share_0_20_prcntl', 'morning_share_20_40_prcntl',
       'morning_share_40_60_prcntl', 'morning_share_60_80_prcntl',
       'morning_share_80_100_prcntl'],
      dtype='object')

In [252]:
interesting_columns =  ['morning_share_0_20_prcntl', 'morning_share_20_40_prcntl',
       'morning_share_60_80_prcntl',
       'morning_share_80_100_prcntl']

In [1]:
stop_words = ['morning', 'category_universities', 'category_colleges']

In [255]:
# Filter the rules
filtered_rules = rules[
    #(rules['antecedents'].apply(lambda x: contains_interesting_column(x, interesting_columns_super_rich)) & 
     #rules['consequents'].apply(lambda x: not contains_stop_word(x, stop_words))) |

    (rules['consequents'].apply(lambda x: contains_interesting_column(x, interesting_columns)) & 
     rules['antecedents'].apply(lambda x: not contains_stop_word(x, stop_words)))
]

# Sort the filtered rules by the lift metric in descending order
filtered_rules_sorted = filtered_rules.sort_values(by='lift', ascending=False)
len(filtered_rules_sorted)

21

In [256]:
usefull_column = ['antecedents', 'consequents',
                 'support', 'confidence', 'lift']

In [257]:
filtered_rules_sorted[filtered_rules_sorted['lift'] > 1][usefull_column].head(10)

antecedents  \
20664                               (category_universities_count_2_hn_binary)   
20668                            (subcategory_universities_count_2_hn_binary)   
20670                                (subcategory_colleges_count_2_hn_binary)   
20666                                   (category_colleges_count_2_hn_binary)   
20676                       (total_subcategory_museums_neighbour_4_hn_binary)   
20678  (total_subcategory_parks_of_culture_and_leisure_neighbour_4_hn_binary)   
14693                                            (age_20_unique_60_80_prcntl)   
12253                                (sprtswr_ftwr_inter_unique_60_80_prcntl)   
14695                                            (age_20_unique_60_80_prcntl)   
11266                                (elite_sports_inter_unique_60_80_prcntl)   

                        consequents   support  confidence      lift  
20664  (morning_share_60_80_prcntl)  0.108614    0.280464  1.402321  
20668  (morning_share_60_80_prcntl)  0.108614    0.280464  1.402321  
20670  (morning_share_60_80_prcntl)  0.108614    0.255732  1.278660  
20666  (morning_share_60_80_prcntl)  0.108614    0.255732  1.278660  
20676  (morning_share_60_80_prcntl)  0.144569    0.237101  1.185504  
20678  (morning_share_60_80_prcntl)  0.167790    0.226950  1.134752  
14693   (morning_share_0_20_prcntl)  0.108614    0.218045  1.090226  
12253   (morning_share_0_20_prcntl)  0.102622    0.216430  1.082148  
14695  (morning_share_20_40_prcntl)  0.103371    0.207519  1.037594  
11266   (morning_share_0_20_prcntl)  0.113109    0.206003  1.030014

## Example of the interpretation for the first case:

In our dataset, the group of visitors aged 20 and below which falls into the 60th to 80th percentile range for unique visits, appears in 50% of the cases (antecedent support = 50%).We see interactions between this age group and salary group 5 (whose unique visit share also falls within the 60th to 80th percentile range). That combination of rules (age_20_unique_60_80_prcntl + salary_5_unique_share_60_80_prcntl) occurs in approximately 11% of cases (support = 11%). The confidence of observing salary group 5's presence in this percentile range given the age group's presence is 22%. It provide 1.10 lift.

- **Antecedent Support (50%)**: tells us that the condition where the age group of 20 and under falls into the 60-80 percentile range for unique visits occurs in half of all observations. This is substantial, indicating a common scenario.

- **Consequent Support (20%)**:  tells that the share of unique visits by representatives of salary group 5 falling into the 60-80 percentile range occurs in 20% of the observations.

- **Support (11%)**: shows combined occurrence (both age and salary conditions met) is seen in 11% of the observations. SHows how often conditions co-occur.

- **Confidence (22%)**: There is a 22% probability that the salary condition is met given that the age condition is already met.

- **Lift (1.10)**: A lift greater than 1 indicates a higher occurrence of this rule than would be expected if the two sets were independent.

**interpretation of choosen columns in percentile ranges with specific values**

In [31]:
path = 'C:/Users/elysi/Tele2_project/data/pre_model_state/df_percentiles_ranges_combined_22_04_2024.csv'
df_percentiles_combined = pd.read_csv(path)

**check all salary columns. so we observe each 20 percentile step and can check clear interpretation of choosen column from 'filtered_rules_sorted' we created before**

In [37]:
df_percentiles_combined[df_percentiles_combined['column_name'].str.contains('salary')]

column_name  0_percentile  20_percentile  40_percentile  \
86   salary_1_unique_share      0.035714       0.140351       0.169766   
87   salary_2_unique_share      0.060000       0.193548       0.224922   
88   salary_3_unique_share      0.066667       0.189129       0.216216   
89   salary_4_unique_share      0.034483       0.151464       0.173554   
90   salary_5_unique_share      0.026316       0.117647       0.145455   
120        salary_1_unique      1.000000      10.000000      15.000000   
121        salary_2_unique      3.000000      14.000000      20.000000   
122        salary_3_unique      2.000000      13.000000      19.000000   
123        salary_4_unique      1.000000      11.000000      16.000000   
124        salary_5_unique      1.000000       9.000000      13.000000   
153   salary_1_total_share      0.006494       0.090710       0.131414   
154   salary_2_total_share      0.022333       0.167211       0.221792   
155   salary_3_total_share      0.012739       0.163154       0.216304   
156   salary_4_total_share      0.017544       0.120643       0.162753   
157   salary_5_total_share      0.005263       0.091563       0.129542   
187         salary_1_total      2.000000      35.000000      63.000000   
188         salary_2_total      7.000000      61.000000     110.600000   
189         salary_3_total      2.000000      59.000000     105.000000   
190         salary_4_total      4.000000      46.000000      83.000000   
191         salary_5_total      2.000000      37.000000      69.000000   

     60_percentile  80_percentile  100_percentile           base  
86        0.193277       0.224138        0.428571  df_hex_unique  
87        0.250000       0.278155        0.468085  df_hex_unique  
88        0.238095       0.264706        0.420000  df_hex_unique  
89        0.193314       0.220669        0.393443  df_hex_unique  
90        0.175000       0.211125        0.416667  df_hex_unique  
120      21.000000      30.000000      131.000000  df_hex_unique  
121      27.000000      37.000000      165.000000  df_hex_unique  
122      27.000000      36.000000      166.000000  df_hex_unique  
123      21.000000      28.000000      135.000000  df_hex_unique  
124      18.000000      25.000000      123.000000  df_hex_unique  
153       0.166978       0.210660        0.570370   df_hex_total  
154       0.268102       0.326107        0.678571   df_hex_total  
155       0.259930       0.315638        0.737624   df_hex_total  
156       0.198948       0.253064        0.609137   df_hex_total  
157       0.171036       0.229405        0.616412   df_hex_total  
187      97.000000     155.000000      568.000000   df_hex_total  
188     164.000000     250.200000      968.000000   df_hex_total  
189     164.000000     249.000000      900.000000   df_hex_total  
190     122.000000     186.000000      732.000000   df_hex_total  
191     102.400000     153.200000      701.000000   df_hex_total

## Apriori with 20 percentile step for each column except salary unique columns.

**Approach presented there combines percentile column of ranges 60-80 and 80-100 into one column**. So it will be easire for apriory to track rules for rich people. 
<br>**Example: In previous case we had:**
1. 'salary_5_unique_share_0_20_prcntl',
2. 'salary_5_unique_share_20_40_prcntl',
3. 'salary_5_unique_share_40_60_prcntl',
4. 'salary_5_unique_share_60_80_prcntl',
5. 'salary_5_unique_share_80_100_prcntl'

**After algorythm we will have:**
1. 'salary_5_unique_share_0_20_prcntl',
2. 'salary_5_unique_share_20_40_prcntl',
3. 'salary_5_unique_share_40_60_prcntl',
4. 'salary_5_unique_share_60_100_prcntl'. 

So we combined salary_5_unique_share_60_80_prcntl and salary_5_unique_share_80_100_prcntl into one column

**Function to make that combination of columns**

In [41]:
binary_df_salary_share_changed = binary_df.copy()

In [42]:
def combine_salary_ranges(df, salary_levels, salary_class):

    for level in salary_levels:
        col_60_80 = f'salary_{level}_{salary_class}_share_60_80_prcntl'
        col_80_100 = f'salary_{level}_{salary_class}_share_80_100_prcntl'
        new_col = f'salary_{level}_{salary_class}_share_60_100_prcntl'

        if col_60_80 in df.columns and col_80_100 in df.columns:
            # logical OR operation
            df[new_col] = df[col_60_80] | df[col_80_100]
            df.drop([col_60_80, col_80_100], axis=1, inplace=True)
    return df

In [43]:
salary_levels = [1, 2, 3, 4, 5]
salary_class = 'unique'

binary_df_salary_share_changed = combine_salary_ranges(binary_df_salary_share_changed, salary_levels, salary_class)

In [44]:
salary_levels = [1, 2, 3, 4, 5]
salary_class = 'total'

binary_df_salary_share_changed = combine_salary_ranges(binary_df_salary_share_changed, salary_levels, salary_class)

In [45]:
salary_share_columns = [col for col in binary_df_salary_share_changed.columns if 'salary' in col.lower() and 'share' in col.lower()]
salary_share_columns

['salary_1_total_share_0_20_prcntl',
 'salary_1_total_share_20_40_prcntl',
 'salary_1_total_share_40_60_prcntl',
 'salary_2_total_share_0_20_prcntl',
 'salary_2_total_share_20_40_prcntl',
 'salary_2_total_share_40_60_prcntl',
 'salary_3_total_share_0_20_prcntl',
 'salary_3_total_share_20_40_prcntl',
 'salary_3_total_share_40_60_prcntl',
 'salary_4_total_share_0_20_prcntl',
 'salary_4_total_share_20_40_prcntl',
 'salary_4_total_share_40_60_prcntl',
 'salary_5_total_share_0_20_prcntl',
 'salary_5_total_share_20_40_prcntl',
 'salary_5_total_share_40_60_prcntl',
 'salary_1_unique_share_0_20_prcntl',
 'salary_1_unique_share_20_40_prcntl',
 'salary_1_unique_share_40_60_prcntl',
 'salary_2_unique_share_0_20_prcntl',
 'salary_2_unique_share_20_40_prcntl',
 'salary_2_unique_share_40_60_prcntl',
 'salary_3_unique_share_0_20_prcntl',
 'salary_3_unique_share_20_40_prcntl',
 'salary_3_unique_share_40_60_prcntl',
 'salary_4_unique_share_0_20_prcntl',
 'salary_4_unique_share_20_40_prcntl',
 'salary_4

## Apriori

In [46]:
%%time

frequent_itemsets = apriori(binary_df_salary_share_changed, min_support=0.1, use_colnames=True, verbose=1, max_len = 2)

rules = association_rules(frequent_itemsets, metric="confidence", min_threshold=0.1)
rules_sorted = rules.sort_values(by='lift', ascending=False)

rules_sorted.head(10)

D:\Anaconda\lib\site-packages\mlxtend\frequent_patterns\fpcommon.py:109: DeprecationWarning: DataFrames with non-bool types result in worse computationalperformance and their support might be discontinued in the future.Please use a DataFrame with bool type
  warnings.warn(


Processing 1027182 combinations | Sampling itemset size 2
CPU times: total: 4.09 s
Wall time: 8.44 s


antecedents  \
23657               (category_free_parking_count_2_hn_20_40_prcntl)   
23656            (subcategory_free_parking_count_2_hn_20_40_prcntl)   
23990     (category_industrial_enterprises_count_2_hn_20_40_prcntl)   
23991  (subcategory_industrial_enterprises_count_2_hn_20_40_prcntl)   
23357                  (subcategory_hotels_count_2_hn_20_40_prcntl)   
23356                     (category_hotels_count_2_hn_20_40_prcntl)   
23781            (subcategory_paid_parking_count_2_hn_20_40_prcntl)   
23780               (category_paid_parking_count_2_hn_20_40_prcntl)   
22392                  (category_hospitals_count_2_hn_40_60_prcntl)   
22393               (subcategory_hospitals_count_2_hn_40_60_prcntl)   

                                                        consequents  \
23657            (subcategory_free_parking_count_2_hn_20_40_prcntl)   
23656               (category_free_parking_count_2_hn_20_40_prcntl)   
23990  (subcategory_industrial_enterprises_count_2_hn_20_40_prcntl)   
23991     (category_industrial_enterprises_count_2_hn_20_40_prcntl)   
23357                     (category_hotels_count_2_hn_20_40_prcntl)   
23356                  (subcategory_hotels_count_2_hn_20_40_prcntl)   
23781               (category_paid_parking_count_2_hn_20_40_prcntl)   
23780            (subcategory_paid_parking_count_2_hn_20_40_prcntl)   
22392               (subcategory_hospitals_count_2_hn_40_60_prcntl)   
22393                  (category_hospitals_count_2_hn_40_60_prcntl)   

       antecedent support  consequent support   support  confidence      lift  \
23657            0.104869            0.104869  0.104869         1.0  9.535714   
23656            0.104869            0.104869  0.104869         1.0  9.535714   
23990            0.110112            0.110112  0.110112         1.0  9.081633   
23991            0.110112            0.110112  0.110112         1.0  9.081633   
23357            0.129588            0.129588  0.129588         1.0  7.716763   
23356            0.129588            0.129588  0.129588         1.0  7.716763   
23781            0.133333            0.133333  0.133333         1.0  7.500000   
23780            0.133333            0.133333  0.133333         1.0  7.500000   
22392            0.147566            0.147566  0.147566         1.0  6.776650   
22393            0.147566            0.147566  0.147566         1.0  6.776650   

       leverage  conviction  zhangs_metric  
23657  0.093871         inf            1.0  
23656  0.093871         inf            1.0  
23990  0.097988         inf            1.0  
23991  0.097988         inf            1.0  
23357  0.112795         inf            1.0  
23356  0.112795         inf            1.0  
23781  0.115556         inf            1.0  
23780  0.115556         inf            1.0  
22392  0.125790         inf            1.0  
22393  0.125790         inf            1.0

**rules for hexagons where share of unique visits by representatives of salary group 4 and 5 lies in percentile 60 - 100 ranges Based on SPB hexagons with total visits lees or equal 150 or unique_visits less or equal 50**

In [47]:
interesting_columns_super_rich = [
 #'salary_4_unique_share_60_100_prcntl',
 'salary_5_unique_share_60_100_prcntl']

In [48]:
filtered_rules = rules[
    #(rules['antecedents'].apply(lambda x: any(item in x for item in interesting_columns_super_rich)) & 
     #rules['consequents'].apply(lambda x: not any("salary" in item for item in x))) |
    
    (rules['consequents'].apply(lambda x: any(item in x for item in interesting_columns_super_rich)) & 
     rules['antecedents'].apply(lambda x: not any("salary" in item for item in x)))
]

# Sort the filtered rules by the lift metric in descending order
filtered_rules_sorted = filtered_rules.sort_values(by='lift', ascending=False)
len(filtered_rules_sorted)

6

In [49]:
filtered_rules_sorted[usefull_column]

antecedents  \
24800                       (total_subcategory_museums_neighbour_4_hn_binary)   
14951                                            (age_20_unique_60_80_prcntl)   
24821  (total_subcategory_parks_of_culture_and_leisure_neighbour_4_hn_binary)   
11704                                (elite_sports_inter_unique_60_80_prcntl)   
24779              (total_subcategory_shopping_centers_neighbour_4_hn_binary)   
24754                       (total_subcategory_cinemas_neighbour_4_hn_binary)   

                                 consequents   support  confidence      lift  
24800  (salary_5_unique_share_60_100_prcntl)  0.137079    0.224816  1.145531  
14951  (salary_5_unique_share_60_100_prcntl)  0.107865    0.216541  1.103369  
24821  (salary_5_unique_share_60_100_prcntl)  0.155056    0.209726  1.068644  
11704  (salary_5_unique_share_60_100_prcntl)  0.111610    0.203274  1.035767  
24779  (salary_5_unique_share_60_100_prcntl)  0.186517    0.194531  0.991218  
24754  (salary_5_unique_share_60_100_prcntl)  0.156554    0.189312  0.964622

**save it as a file**

In [79]:
rules_sorted.shape

(24832, 10)

In [72]:
path = 'C:/Users/elysi/Tele2_project/data/pre_model_state/ARL_salary_60_100.csv'
rules_sorted.to_csv(path, index=False)

## FP Grows

**for FP Grows we need boolean dataset since it works with optimization beter**

In [17]:
binary_df_bool = binary_df.astype(bool)

In [19]:
binary_df_bool.shape

(1394, 1036)

In [20]:
binary_df_bool.head(3)

total_visits_0_20_prcntl  total_visits_20_40_prcntl  \
h3_9                                                                   
89110600113ffff                      True                      False   
8911060018fffff                     False                       True   
891106004b7ffff                     False                      False   

                 total_visits_40_60_prcntl  total_visits_60_80_prcntl  \
h3_9                                                                    
89110600113ffff                      False                      False   
8911060018fffff                      False                      False   
891106004b7ffff                      False                       True   

                 total_visits_80_100_prcntl  car_owners_total_0_20_prcntl  \
h3_9                                                                        
89110600113ffff                       False                          True   
8911060018fffff                       False                         False   
891106004b7ffff                       False                         False   

                 car_owners_total_20_40_prcntl  car_owners_total_40_60_prcntl  \
h3_9                                                                            
89110600113ffff                          False                          False   
8911060018fffff                           True                          False   
891106004b7ffff                          False                          False   

                 car_owners_total_60_80_prcntl  \
h3_9                                             
89110600113ffff                          False   
8911060018fffff                          False   
891106004b7ffff                           True   

                 car_owners_total_80_100_prcntl  \
h3_9                                              
89110600113ffff                           False   
8911060018fffff                           False   
891106004b7ffff                           False   

                 carsharing_user_total_0_20_prcntl  \
h3_9                                                 
89110600113ffff                               True   
8911060018fffff                               True   
891106004b7ffff                              False   

                 carsharing_user_total_20_40_prcntl  \
h3_9                                                  
89110600113ffff                               False   
8911060018fffff                               False   
891106004b7ffff                               False   

                 carsharing_user_total_40_60_prcntl  \
h3_9                                                  
89110600113ffff                               False   
8911060018fffff                               False   
891106004b7ffff                               False   

                 carsharing_user_total_60_80_prcntl  \
h3_9                                                  
89110600113ffff                               False   
8911060018fffff                               False   
891106004b7ffff                               False   

                 carsharing_user_total_80_100_prcntl  \
h3_9                                                   
89110600113ffff                                False   
8911060018fffff                                False   
891106004b7ffff                                False   

                 not_citizen_rf_total_0_20_prcntl  \
h3_9                                                
89110600113ffff                              True   
8911060018fffff                             False   
891106004b7ffff                             False   

                 not_citizen_rf_total_20_40_prcntl  \
h3_9                                                 
89110600113ffff                              False   
8911060018fffff                               True   
891106004b7ffff                              False   

                 not_citizen_rf_total_40_60_prcntl  \
h3_9                

In [23]:
%%time
from mlxtend.frequent_patterns import fpgrowth, association_rules
frequent_itemsets_fp = fpgrowth(binary_df_bool, min_support=0.1, use_colnames=True, max_len=2)

rules_fp = association_rules(frequent_itemsets_fp, metric="confidence", min_threshold=0.3)

rules_sorted_fp = rules_fp.sort_values(by='lift', ascending=False)

rules_sorted_fp.head(10)

CPU times: total: 8min 28s
Wall time: 11min 51s


antecedents                   consequents  \
57419    (female_total_share_60_80)      (male_total_share_20_40)   
57418      (male_total_share_20_40)    (female_total_share_60_80)   
24535   (female_unique_share_20_40)     (male_unique_share_60_80)   
24536     (male_unique_share_60_80)   (female_unique_share_20_40)   
19010   (unique_clubs_names_binary)      (total_fit_clubs_binary)   
22883  (unique_clubs_sortes_binary)      (total_fit_clubs_binary)   
22882      (total_fit_clubs_binary)  (unique_clubs_sortes_binary)   
22881  (unique_clubs_sortes_binary)   (unique_clubs_names_binary)   
22880   (unique_clubs_names_binary)  (unique_clubs_sortes_binary)   
19009      (total_fit_clubs_binary)   (unique_clubs_names_binary)   

       antecedent support  consequent support   support  confidence      lift  \
57419            0.102582            0.103300  0.101865    0.993007  9.612859   
57418            0.103300            0.102582  0.101865    0.986111  9.612859   
24535            0.167862            0.160689  0.160689    0.957265  5.957265   
24536            0.160689            0.167862  0.160689    1.000000  5.957265   
19010            0.203730            0.203730  0.203730    1.000000  4.908451   
22883            0.191535            0.203730  0.191535    1.000000  4.908451   
22882            0.203730            0.191535  0.191535    0.940141  4.908451   
22881            0.191535            0.203730  0.191535    1.000000  4.908451   
22880            0.203730            0.191535  0.191535    0.940141  4.908451   
19009            0.203730            0.203730  0.203730    1.000000  4.908451   

       leverage  conviction  zhangs_metric  
57419  0.091268  128.228121       0.998390  
57418  0.091268   64.614060       0.999189  
24535  0.133715   19.639885       1.000000  
24536  0.133715         inf       0.991453  
19010  0.162224         inf       1.000000  
22883  0.152514         inf       0.984916  
22882  0.152514   13.506119       1.000000  
22881  0.152514         inf       0.984916  
22880  0.152514   13.506119       1.000000  
19009  0.162224         inf       1.000000

**works the same, we can use it for optimized runs. In theory it provided opportunity for 3 rules combination. Nevertheless, currently we work we pairs of rules which apriori algorythm can handle**